aggregator/handler/config.go

```diff
@@ -137,10 +135,98 @@ func (c *forwardHandlerConfiguration) NewHandler(
 	if c.ConnectionWriteTimeout != 0 {
 		opts = opts.SetConnectionWriteTimeout(c.ConnectionWriteTimeout)
 	}
-
-	scope := instrumentOpts.MetricsScope().SubScope("reconnect")
-	retrier := c.ReconnectRetrier.NewRetrier(scope)
+	reconnectScope := scope.SubScope("reconnect")
+	retrier := c.ReconnectRetrier.NewRetrier(reconnectScope)
 	opts = opts.SetReconnectRetrier(retrier)
+	return opts
+}
+
+type partitionedHandlerConfiguration struct {
+	// Name of the partitioned backend.
+	Name string `yaml:"name"`
+
+	// Backend server partitions.
+	Partitions []BackendServerPartition
+
+	// Total queue size across all partitions.
+	QueueSize int `yaml:"queueSize"`
+
+	// Connection configuration.
+	Connection connectionConfiguration `yaml:"connection"`
+}
+
+func (c *partitionedHandlerConfiguration) NewHandler(
+	instrumentOpts instrument.Options,
+) (aggregator.Handler, error) {
+	if err := c.validatePartitions(); err != nil {
+		return nil, err
+	}
+	connectionOpts := c.Connection.NewConnectionOptions(instrumentOpts.MetricsScope())
+	opts := NewOptions().
+		SetInstrumentOptions(instrumentOpts).
+		SetConnectionOptions(connectionOpts)
+	if c.QueueSize != 0 {
+		opts = opts.SetQueueSize(c.QueueSize)
+	}
+	return NewPartitionedHandler(c.Partitions, opts)
+}
+
+// validatePartitions ensures a single partition isn't present multiple times, and
+// that a single server isn't assigned to multiple partition ranges.
+func (c *partitionedHandlerConfiguration) validatePartitions() error {
+	var (
+		serversAssigned    = make(map[string]struct{})
+		partitionsAssigned = make(map[int]struct{})
+	)
+	for _, partitions := range c.Partitions {
+		// Make sure we have a deterministic ordering.
+		sortedPartitions := make([]int, 0, len(partitions.Range))
+		for partition := range partitions.Range {
+			sortedPartitions = append(sortedPartitions, int(partition))
+		}
+		sort.Ints(sortedPartitions)
```

Comment:
Any reason not to implement `sort.Interface` for `BackendServerPartition` so we can sort `partitions` directly?

Reply:
Not sure what you meant by "sort partitions directly", internally the `backendServerPartition` has a partition set which is a map[uint32]struct{}, this is to sort the partitions contained in the set to have deterministic ordering when validating them.

Reply:
Ahh, disregard, I thought `partitions.Range` was a slice for some reason.